### PR TITLE
Fix/revisit git-describe enhancement

### PR DESCRIPTION
**Pull request by blueyed**
1. Fix/keep massaging git-describe result

    Ref: https://github.com/neovim/neovim/pull/11117#issuecomment-536416223

TODO:

- [x] no git-describe output during configure (would need to be massaged also, but is not necessary after all)
- [x] `make clean && make` is broken (since the BYPRODUCT apparently gets removed then..)

### CODE REVIEW

**Comment by h-michael** [comment]
On linux

current head.
```
$ nvim --version
NVIM v0.4.0-99-gb18b84df5
Build type: Release
```

this pr's merge commit.
```
$ ./build/bin/nvim --version
NVIM v0.5.0-103-gf1d734030
Build type: Release
```

**Comment by blueyed** [comment]
For reference, on Sourcehut:
```
[15/323] cd /usr/home/build/neovim && /usr/home/build/neovim/.deps/usr/bin/luajit scripts/update_version_stamp.lua build/config/auto/versiondef_git.h v0.5.0
git: 'describee' is not a git command. See 'git --help'.

The most similar command is
	describe
git-describe failed
debug: --always	729c2532e
debug: --tags	nightly-12207-g729c2532e
debug: --tags --always	nightly-12207-g729c2532e
```

I've used `--tags --always` as fallback now.

**Comment by h-michael** [comment]
On MacOS

This is current HEAD.
```
$ ~> brew install --HEAD neovim
==> Cloning https://github.com/neovim/neovim.git
Updating /Users/foo/Library/Caches/Homebrew/neovim--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at b18b84df5 build: run git-describe for dev version during build (#11117)
==> Downloading https://github.com/libmpack/libmpack-lua/releases/download/1.0.7/libmpack-lua-1.0.7.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/ad0f976ceb0dc87d6ce2195e0a405fc851ae85b558580477f1bac0c93c2a933a--libmpack-lua-1.0.7.tar.gz
==> Downloading https://luarocks.org/manifests/gvvaughan/lpeg-1.0.2-1.src.rock
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/0bc17975f0cf3981083172aa7da329b874e5780fa81494f5dd07d0665fcec23b--lpeg-1.0.2-1.src.rock
On MacOS

This is current HEAD.
```
$ ~> brew install --HEAD neovim
==> Cloning https://github.com/neovim/neovim.git
Updating /Users/foo/Library/Caches/Homebrew/neovim--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at b18b84df5 build: run git-describe for dev version during build (#11117)
==> Downloading https://github.com/libmpack/libmpack-lua/releases/download/1.0.7/libmpack-lua-1.0.7.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/ad0f976ceb0dc87d6ce2195e0a405fc851ae85b558580477f1bac0c93c2a933a--libmpack-lua-1.0.7.tar.gz
==> Downloading https://luarocks.org/manifests/gvvaughan/lpeg-1.0.2-1.src.rock
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/0bc17975f0cf3981083172aa7da329b874e5780fa81494f5dd07d0665fcec23b--lpeg-1.0.2-1.src.rock
==> Downloading https://luarocks.org/manifests/kikito/inspect-3.1.1-0.src.rock
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/ecc1fa84ce4c83c9406f6d1202d4a88c80ede4a0a2bfd8d8d2bd54fd042226bb--inspect-3.1.1-0.src.rock
==> Downloading https://github.com/keplerproject/lua-compat-5.3/archive/v0.7.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/5f4e552e95b9a6223a5e4fabb7f04bf4b886057024507c76b1d5b803218c9363--lua-compat-5.3-0.7.tar.gz
==> Downloading https://github.com/luvit/luv/releases/download/1.30.0-0/luv-1.30.0-0.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/a9ce5a949370965d8f0fe70a47e7d102a3b4b6b831dbd4d088a83bbe1c00a3a2--luv-1.30.0-0.tar.gz
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-43531-1fngf29/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-43531-1fngf29/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-43531-1fngf29/deps-build
==> cmake . -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DCMAKE_
==> make install
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/neovim/HEAD-b18b84d -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIN
Last 15 lines from /Users/foo/Library/Logs/Homebrew/neovim/06.cmake:
    UTF8PROC_INCLUDE_DIR=<not found>
    UTF8PROC_LIBRARY=<not found>

  You may use CMake GUI, cmake -D or ccmake to modify the values.  Delete
  CMakeCache.txt to discard all values and force full re-detection if
  necessary.

Call Stack (most recent call first):
  cmake/FindUTF8PROC.cmake:16 (libfind_process)
  CMakeLists.txt:365 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/neovim-20190930-43531-1fngf29/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/neovim-20190930-43531-1fngf29/build/CMakeFiles/CMakeError.log".

READ THIS: https://docs.brew.sh/Troubleshooting
```

```
$ ~> brew config
HOMEBREW_VERSION: 2.1.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 37714b5ce1e4909d4d61de2af98343c7012f7cd9
Last commit: 5 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: af98c4bac2755d282b976d55dc995dd9cba9f4de
Core tap last commit: 2 hours ago
HOMEBREW_PREFIX: /usr/local
CPU: octa-core 64-bit kabylake
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1001
Git: 2.23.0 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.14.6-x86_64
CLT: 10.2.1.0.1.1554506761
Xcode: 10.3
CLT headers: 10.2.1.0.1.1554506761
``

**Comment by blueyed** [comment]
@h-michael 
The last one is due to missing UTF8PROC (not related to the fixes here).

**Comment by blueyed** [comment]
> So we may update homebrew cask.

Do you mean "task"?

Anyway, this PR is about fixing the `nvim --version` regression from b18b84d.

**Comment by h-michael** [comment]
This is current HEAD by homebrew.

```
$ ~> brew install utf8proc
Updating Homebrew...
==> Downloading https://homebrew.bintray.com/bottles/utf8proc-2.4.0.mojave.bottle.tar.gz
######################################################################## 100.0%
==> Pouring utf8proc-2.4.0.mojave.bottle.tar.gz
🍺  /usr/local/Cellar/utf8proc/2.4.0: 10 files, 644.2KB
$ ~> brew install --HEAD neovim
Updating Homebrew...
==> Cloning https://github.com/neovim/neovim.git
Updating /Users/foo/Library/Caches/Homebrew/neovim--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at b18b84df5 build: run git-describe for dev version during build (#11117)
==> Downloading https://github.com/libmpack/libmpack-lua/releases/download/1.0.7/libmpack-lua-1.0.7.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/ad0f976ceb0dc87d6ce2195e0a405fc851ae85b558580477f1bac0c93c2a933a--libmpack-lua-1.0.7.tar.gz
==> Downloading https://luarocks.org/manifests/gvvaughan/lpeg-1.0.2-1.src.rock
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/0bc17975f0cf3981083172aa7da329b874e5780fa81494f5dd07d0665fcec23b--lpeg-1.0.2-1.src.rock
==> Downloading https://luarocks.org/manifests/kikito/inspect-3.1.1-0.src.rock
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/ecc1fa84ce4c83c9406f6d1202d4a88c80ede4a0a2bfd8d8d2bd54fd042226bb--inspect-3.1.1-0.src.rock
==> Downloading https://github.com/keplerproject/lua-compat-5.3/archive/v0.7.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/5f4e552e95b9a6223a5e4fabb7f04bf4b886057024507c76b1d5b803218c9363--lua-compat-5.3-0.7.tar.gz
==> Downloading https://github.com/luvit/luv/releases/download/1.30.0-0/luv-1.30.0-0.tar.gz
Already downloaded: /Users/foo/Library/Caches/Homebrew/downloads/a9ce5a949370965d8f0fe70a47e7d102a3b4b6b831dbd4d088a83bbe1c00a3a2--luv-1.30.0-0.tar.gz
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-56956-i8gdj7/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-56956-i8gdj7/deps-build
==> luarocks make --lua-dir=/usr/local/opt/luajit --tree=/private/tmp/neovim-20190930-56956-i8gdj7/deps-build
==> cmake . -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DCMAKE_INSTALL_PREFIX=/private/
==> make install
==> cmake .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/neovim/HEAD-b18b84d -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE
==> make install
🍺  /usr/local/Cellar/neovim/HEAD-b18b84d: 1,422 files, 19.8MB, built in 6 minutes 2 seconds
$ ~> nvim --version
NVIM v0.5.0-dev
Build type: Release
```

**Comment by blueyed** [comment]
@h-michael 
> NVIM v0.5.0-dev

Hmm..  can you post the build log?

**Comment by clason** [comment]
That (0.5.0-dev) also happens on one(!) of my macOS machines, not sure why. It’s the newest (i.e., the one where homebrew and neovim was installed most recently), if that makes any difference.

**Comment by blueyed** [comment]
Maybe also built with a bad commit from this PR?

I've just tested `rm build -rf && make` on master, followed by checking out that branch and just `make` again, resulting in `v0.5.0-106-gc038ee27f` (expected).

**Comment by blueyed** [comment]
"0.5.0-dev" would indicate that Git does not work..

Can you post the output of `find -name '*versiondef*'`, and `grep . ./build/config/auto/versiondef*`?

**Comment by h-michael** [comment]
@blueyed 
This is build log on MacOS by using make instead of homebrew.
https://gist.github.com/h-michael/199162ebeeec3acf97ed8f5fba99ce00

**Comment by h-michael** [comment]
@blueyed 
This is a build log which build this PR on MacOS.
https://gist.github.com/h-michael/803f4de0067108aa9f8ea3800f533885

**Comment by blueyed** [comment]
@h-michael 
Thanks, that looks good:

> -- Using NVIM_VERSION_MEDIUM from Git
> NVIM v0.5.0-106-gc038ee27f

@clason 
Can you provide more insight (log) for the "0.5.0-dev" case?

**Comment by clason** [comment]
@blueyed I cannot, sorry; in the configure log I see
```
fatal: No names found, cannot describe anything.
-- NVIM_VERSION_MEDIUM (from git):
```
but I have no idea why this happens on this and only this machine. `git version 2.23.0` is installed, of course.

**Comment by blueyed** [comment]
@clason 
Does `git describe --dirty` work in the source root?
(if not, does `git describe --tags --always --dirty`?)
Are you using the latest commit from this PR/branch?

**Comment by clason** [comment]
@blueyed Hmm, first of all I have to find where homebrew stores the local repo...

**Comment by blueyed** [comment]
Ah, thought you were using this PR.

I think this is good then to be merged.  It should fix your issue.

File: src/nvim/CMakeLists.txt [hunk 1]
```diff
@@ -217,6 +217,34 @@ function(get_preproc_output varname iname)
...
+# Handle generating version from Git.
+set(use_git_version 0)
+if(NVIM_VERSION_MEDIUM)
+  message(STATUS "NVIM_VERSION_MEDIUM: ${NVIM_VERSION_MEDIUM}")
+elseif(${CMAKE_VERSION} VERSION_LESS "3.2.0")
+  message(STATUS "Skipping version-string generation (requires CMake 3.2.0+)")
+elseif(EXISTS ${PROJECT_SOURCE_DIR}/.git)
+  find_program(GIT_EXECUTABLE git)
+  if(GIT_EXECUTABLE)
+    message(STATUS "Using NVIM_VERSION_MEDIUM from Git")
```
Comment (blueyed) [inline src/nvim/CMakeLists.txt]: This could maybe still call the script in a special way to get the output that would be used.
But it's just cosmetics.

**Review by justinmk — APPROVED** [review]

**Comment by camponez** [comment]
Building  b18b84df5eab9829ec^ (previous commit)

```
$ ./bin/nvim --version
NVIM v0.5.0-98-gce637d0ef
``` 

Build b18b84df5eab9829ec

```
-- NVIM_VERSION_MEDIUM (from git): v0.4.0-99-gb18b84df5
```

**Comment by blueyed** [comment]
@camponez 
But ok with 30ae60e, no?

**Comment by camponez** [comment]
It was broken on 8a4ae3d66 (yesterday's nightly)

but ok again on 0c1be45ea (today's nightly)

cheers.

**Pull request merged**